tools/src/org.graalvm.tools.lsp/src/org/graalvm/tools/lsp/server/utils/TextDocumentSurrogate.java

```diff
@@ -0,0 +1,200 @@
+package org.graalvm.tools.lsp.server.utils;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.eclipse.lsp4j.TextDocumentContentChangeEvent;
+
+import com.oracle.truffle.api.CallTarget;
+import com.oracle.truffle.api.nodes.LanguageInfo;
+import com.oracle.truffle.api.source.Source;
+import com.oracle.truffle.api.source.SourceSection;
+
+public final class TextDocumentSurrogate {
```

Comment:
Does this needs thread-safety?

Reply:
Not yet, because the current implementation uses only a single context-entered worker thread which accesses instances of this class.